src/nnet3/nnet-simple-component.cc

```diff
@@ -633,6 +633,135 @@ void NoOpComponent::Backprop(const std::string &debug_info,
+  BaseFloat regularization_scale = 0.1;
+  cfl->GetValue("regularization-scale", &regularization_scale);
+  if (!ok || cfl->HasUnusedValues() || regularization_scale < 0.0 || dim <= 0)
+    KALDI_ERR << "Invalid initializer for layer of type "
+              << Type() << ": \"" << cfl->WholeLine() << "\"";
+  Init(dim, regularization_scale);
+}
+
+
+void SpatialRegularizationComponent::Init(int32 dim, BaseFloat regularization_scale) {
+  KALDI_ASSERT(dim > 0 && regularization_scale >= 0.0);
+  dim_ = dim;
+  regularization_scale_ = regularization_scale;
+  count_ = 0.0;
+  sumsq_ = 0.0;
+}
+
+void SpatialRegularizationComponent::Read(std::istream &is, bool binary) {
+  ExpectOneOrTwoTokens(is, binary, "<SpatialRegularizationComponent>",
+                       "<Dim>");
+  ReadBasicType(is, binary, &dim_);
+  ExpectToken(is, binary, "<RegularizationScale>");
+  ReadBasicType(is, binary, &regularization_scale_);
+  ExpectToken(is, binary, "<Count>");
+  ReadBasicType(is, binary, &count_);
+  ExpectToken(is, binary, "<Sumsq>");
+  ReadBasicType(is, binary, &sumsq_);
+  ExpectToken(is, binary, "</SpatialRegularizationComponent>");
+}
+
+
+void SpatialRegularizationComponent::Write(std::ostream &os,
+                                           bool binary) const {
+  WriteToken(os, binary, "<SpatialRegularizationComponent>");
+  WriteToken(os, binary, "<Dim>");
+  WriteBasicType(os, binary, dim_);
+  WriteToken(os, binary, "<RegularizationScale>");
+  WriteBasicType(os, binary, regularization_scale_);
+  WriteToken(os, binary, "<Count>");
+  WriteBasicType(os, binary, count_);
+  WriteToken(os, binary, "<Sumsq>");
+  WriteBasicType(os, binary, sumsq_);
+  WriteToken(os, binary, "</SpatialRegularizationComponent>");
+}
+
+void SpatialRegularizationComponent::Scale(BaseFloat scale) {
+  count_ *= scale;
+  sumsq_ *= scale;
+}
+
+void SpatialRegularizationComponent::Add(BaseFloat alpha,
+                                         const Component &other_in) {
+  const SpatialRegularizationComponent *other =
+      dynamic_cast<const SpatialRegularizationComponent*>(&other_in);
+  KALDI_ASSERT(other != NULL);
+  count_ += alpha * other->count_;
+  sumsq_ += alpha * other->sumsq_;
+}
+
+void SpatialRegularizationComponent::Propagate(
+    const ComponentPrecomputedIndexes *indexes,
+    const CuMatrixBase<BaseFloat> &in,
+    CuMatrixBase<BaseFloat> *out) const {
+  if (out != &in)
+    out->CopyFromMat(in);
+}
+
+void SpatialRegularizationComponent::Backprop(
+    const std::string &debug_info,
+    const ComponentPrecomputedIndexes *indexes,
+    const CuMatrixBase<BaseFloat> &, //in_value
+    const CuMatrixBase<BaseFloat> &out_value,
+    const CuMatrixBase<BaseFloat> &out_deriv,
+    Component *to_update_in,
+    CuMatrixBase<BaseFloat> *in_deriv) const {
+  if (in_deriv != &out_deriv)
+    in_deriv->CopyFromMat(out_deriv);
+  // accumulate stats only every other time.
+  bool compute_sumsq = (to_update_in != NULL && RandInt(0, 1) == 0);
```

Comment:
You can change this to RandInt(0, 3) == 0, since computing the sumsq is a little slow.  [this is only for diagnostics.]